centralstore/redis_store.go

```diff
@@ -900,77 +894,108 @@ func (t *traceStateProcessor) removeExpiredTraces(client redis.Client) {
 
 }
 
-// get the latest state from the state list
-// check if the current state change event is valid
-// if it is, then add the state to state list
-// and add the trace to the state set
-// if it's not, then don't add the state to the state list	and abort
-func (t *traceStateProcessor) applyStateChange(ctx context.Context, conn redis.Conn, stateChange stateChangeEvent, traceID string) error {
+// applyStateChange runs a lua script that atomically moves traces between states and returns the trace IDs that has completed a state change.
+func (t *traceStateProcessor) applyStateChange(ctx context.Context, conn redis.Conn, stateChange stateChangeEvent, traceIDs []string) ([]string, error) {
 	ctx, span := t.tracer.Start(ctx, "applyStateChange")
 	defer span.End()
 
-	otelutil.AddSpanField(span, "trace_id", traceID)
+	otelutil.AddSpanField(span, "num_traces", len(traceIDs))
+
+	args := redigo.Args{validStateChangeEventsKey, stateChange.current.String(), stateChange.next.String(),
```

Comment:
TODO: I need to think about how to extract this redigo stuffs back into our internal redis package

Reply:
Maybe make a helper function args() that returns a redigo.Args?